### PR TITLE
[tf.image.convert_image_dtype] - Add extra warning if scaling is skipped

### PR DESCRIPTION
**Pull request by DEKHTIARJonathan**
We have noticed recently a number of people facing input data scaling using `tf.image.convert_image_dtype`. 

The problem is the following: we see users converting a tensor dtype unwantedly and having the normalization step skipped silently. See example below:

```python
# Bad order of operations
img_file = tf.read_file("my_filename_here.png")   # Binary dtype (string)
img = tf.image.decode_png(img_file, channels=3, dtype=tf.uint8)   # UINT8 dtype
img = tf.image.resize_images(img, shape, shape)   # Silently converts to FP32
img = tf.convert_image_dtype(img , tf.float32)   # this operation is skipped silently

# Correct order of operations
img_file = tf.read_file("my_filename_here.png")   # Binary dtype (string)
img = tf.image.decode_png(img_file, channels=3, dtype=tf.uint8)   # UINT8 dtype
img = tf.convert_image_dtype(img , tf.float32)   # Converted to FP32 + Normalized
img = tf.image.resize_images(img, shape, shape)   # Rescaling
```

The objective here is really simple, making sure the user is aware of this behavior and realize that the expected behavior might not be respected.
In addition the documentation should IMO highlight that the normalization is not systematic and only takes place when going from INT format to FP format.

CC: @reedwm

### CODE REVIEW

**Comment by martinwicke** [comment]
I actually think this is a bug (borderline) in resize_images. I would prefer if resize_images warned about automatically casting (but not properly converting) an image. I.e., if it warned about integer inputs. 

I do have a performance concern about a warning in these functions, they are often in input pipelines, and if we add a warning, it should be in C++ (inside the resize kernels which do the actual conversion), and it should be smart about not warning for every image, since most likely this affects either every image or none.

**Comment by DEKHTIARJonathan** [comment]
@martinwicke 

> I do have a performance concern about a warning in these functions, they are often in input pipelines, and if we add a warning, it should be in C++ (inside the resize kernels which do the actual conversion), and it should be smart about not warning for every image, since most likely this affects either every image or none.

I don't really agree here. The warning wouldn't be apply at each iterations. Only when you are designing the dataloading graph. The cost of one warning is virtually none.

> I actually think this is a bug (borderline) in resize_images

Even if we fix it, that does not fix the fact that if you input a floating point tensor (fp16, or fp32). I feel like the documentation is unclear and it leads a lot of users to crash into issues that are expensive to debug (finding that the rescale you expected is not here is no trivial task).

Just in a one minute google search, we can find example of people bumping into this issue:
https://stackoverflow.com/questions/54972167/bug-with-tf-image-convert-image-dtype-values-not-scaling

**Comment by martinwicke** [comment]
If you're strictly talking about the compat.v1 version I can see what you
mean. But remember that TensorFlow supports (preferentially) eager
execution, so the Python function will be executed for each batch.

I know that lots of users are confused. And if you look at other packages
(e.g. OpenCV), you'll see that they have taken pretty similar approaches.
For better or worse, image representations are underspecified, and in order
to write processing functions, a standard is necessary. The one chosen in
TF is: float images are assumed to be [0,1) normalized (unless HDR), and
integer images are [0,max] representing fix point [0,1).

The bug in resize I am referring to is that it accepts non-float images,
but does not renornalize them. It therefore can return float images which
are not in a native format.

I don't believe we can change this behavior, but we can warn about integer
images passed directly to resize.

Independently, I agree this isn't well-understood, and If you have a
suggestion on how to improve the documentation I am definitely interested.

>

**Comment by gbaned** [comment]
@DEKHTIARJonathan Can you please check @martinwicke's comments and keep us posted. Thanks!

**Comment by DEKHTIARJonathan** [comment]
@martinwicke sorry for the delay in answer.

Implementing a warning in `tf.image.resize_image` or more precisely this function: https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/ops/image_ops_impl.py#L1180-L1255 sounds a good idea. However, it can't be added in C++ without a massive refactoring. Most of the code is Python logic relying on basic/common TF C++ Operations.

What I can do is to add a warning and make sure it's only printed once (probably using global var, I will see if I can avoid it).

-------

For the second point, `tf.convert_image_dtype` should still IMHO issue a warning the user requests same kind of inputs/outputs dtype (integers/floating). Resize is just one scenario triggering this issue, it could happen with any TF operation that changes the dtype without the user realizing. Unfortunately same issue here again, no C++ code, just python logic and basic/common C++ TF Operations (identity and cast for our usecase). We can nonetheless apply the same trick as for `tf.image.resize_image` to warn only once the user.

**Comment by martinwicke** [comment]
Fair point. Looking at the size of the Python logic in that function, it looks to me like we can add a simple `if input_dtype == output_dtype: WARN` without fear of terrible performance implications. 

We have code to limit the warnings to the first n in platform/tf_logging, but we have deprecated that code since it's too close to generic python logging. Fundamentally, any such limit would involve a global (and that's a fine use of a global in my opinion).

However, that is a question that @tensorflow/api-owners should maybe be involved in.

**Comment by DEKHTIARJonathan** [comment]
I believe there is two potential path dangerous for this API:
- identical dtypes (you already pointed out this one).
- floating to floating: with more and more integration of FP16/BFLOAT. It would be wise to also protect that route. For some reason an OP could convert to FP16/BFLOAT16 without informing the user.

@tensorflow/api-owners Please let me know how you would like this bug fixed. Or feel free to directly correct it ;)

**Comment by martinwicke** [comment]
Ah. Sorry, I mistyped this: I meant to say, we should add the following check to `tf.image.resize_image` (and other endpoints which share the "feature" of potentially silently changing dtypes):

```
if output_dtype != input_dtype: WARN
```

**Review by mihaimaruseac — DISMISSED** [review]

**Review by martinwicke — CHANGES_REQUESTED** [review]
The discussion isn't done just yet.

**Comment by martinwicke** [comment]
Np. Just left a review requesting changes so status is clear.

**Comment by gbaned** [comment]
@Jonathan DEKHTIAR Any update on this PR? Please. Thanks!

**Comment by martinwicke** [comment]
I think the comments in your code snippet are wrong?

**Comment by DEKHTIARJonathan** [comment]
@martinwicke oh yes my bad. Fixing immediately the post above

File: tensorflow/python/ops/image_ops_impl.py [hunk 1]
```diff
@@ -1179,6 +1180,46 @@ class ResizeMethod(object):
   MITCHELLCUBIC = 'mitchellcubic'
 
 
+def _check_equal_input_output_dtypes(func):
+  """This decorator issue will issue a warning if the input dtype and output
+  dtype are different. Help to prevent silent dtype conversion from integer to
+  floating point format."""
+
+  def wrapper(*args, **kwargs):
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_impl.py:1188]: A decorator is very ugly in stack traces, can we make this into a utility function which is called from `_resize_images_common` instead?

File: tensorflow/python/ops/image_ops_impl.py [hunk 2]
```diff
@@ -1179,6 +1180,46 @@ class ResizeMethod(object):
...
+    try:
+      check_dtype = _check_equal_input_output_dtypes._fns_flagged[func.__name__]
+    except KeyError:
+      _check_equal_input_output_dtypes._fns_flagged[func.__name__] = True
+      check_dtype = True
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_impl.py:1193]: Can we change this to the simpler (and more efficient, at least after we made this into a utility function):

```
if input_dtype == output_dtype:
  return

if funcname in _CHECK_EQUAL_INPUT_OUTPUT_FNS_FLAGGED:
  return

# print warning
_CHECK_EQUAL_INPUT_OUTPUT_FNS_FLAGGED.add(funcname)


```

File: tensorflow/python/ops/image_ops_impl.py [hunk 3]
```diff
@@ -1179,6 +1180,46 @@ class ResizeMethod(object):
...
+    if check_dtype and (input_dtype != output.dtype):
+        _check_equal_input_output_dtypes._fns_flagged[func.__name__] = False
+        logging.warning(
+          "The operation `{func_name}` has silently converted the "
+          "data type from `{input_dtype}` to `{output_dtype}`. "
+          "This might have an adverse effect on data numerical "
+          "range.".format(
+              func_name=func.__name__,
+              input_dtype=input_dtype,
+              output_dtype=output.dtype
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_impl.py:1212]: Can we make this print the line number of the call to the function which triggered the warning? That would mean looking back in the stacktrace a few frames up ("a few" meaning how every many needed. If this utility function is used by different functions (eventually) the number of frames to unroll for this might have to be a parameter).

File: tensorflow/python/ops/image_ops_impl.py [hunk 4]
```diff
@@ -1179,6 +1180,46 @@ class ResizeMethod(object):
...
+
+    output = func(*args, **kwargs)
+
+    if check_dtype and (input_dtype != output.dtype):
+        _check_equal_input_output_dtypes._fns_flagged[func.__name__] = False
+        logging.warning(
+          "The operation `{func_name}` has silently converted the "
+          "data type from `{input_dtype}` to `{output_dtype}`. "
+          "This might have an adverse effect on data numerical "
+          "range.".format(
```
Comment (martinwicke) [inline tensorflow/python/ops/image_ops_impl.py:1209]: I would rephrase to: 

"This might have changed the image format in ways that make it incompatible with other `tf.image` functions."

**Review by martinwicke — CHANGES_REQUESTED** [review]

**Comment by martinwicke** [comment]
Cool. Some stylistic comments, I think in principle this will work.

**Comment by gbaned** [comment]
@DEKHTIARJonathan Can you please check @martinwicke's comments and keep us posted. Thanks!

**Comment by gbaned** [comment]
@DEKHTIARJonathan Any update on this PR? Please. Thanks!

File: tensorflow/python/ops/image_ops_impl.py (hunk 1 above)
Comment (DEKHTIARJonathan) [reply tensorflow/python/ops/image_ops_impl.py:1214]: Not easily doable. We need to access the input and output dtypes to compare them.
Might not be ideal but this is really a case for decorators... Aside of this I would prefer not touching the logic of these functions.
If you really don't want a decorator, can you be more specific on the solution you have in mind ?

**Review by DEKHTIARJonathan — COMMENTED** [review]
@martinwicke changes requested implemented in commit 6e1647890050102834754e645d5a9ed2b9d35734

Which now gives the following output:
```python
import os
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"

import tensorflow as tf


if __name__ == "__main__":

    print("Processing with the correct order of operations ... "
          "No warning shall be printed ...")
    for step in range(100):
        # Bad order of operations
        img_file = tf.compat.v1.read_file("my_image.png")   # Binary dtype (string)
        img = tf.image.decode_png(img_file, channels=3, dtype=tf.uint8)   # UINT8 dtype
        img = tf.image.convert_image_dtype(img, tf.float32)   # Converts explicitly to FP32
        img = tf.image.resize(img, (255, 255))   # No silent  conversion happens as the input is already in FP32

    print("\nProcessing with the dangerous order of operations ... "
          "Only one warning shall be printed ...")
    for step in range(100):
        # Bad order of operations
        img_file = tf.compat.v1.read_file("my_image.png")   # Binary dtype (string)
        img = tf.image.decode_png(img_file, channels=3, dtype=tf.uint8)   # UINT8 dtype
        img = tf.image.resize(img, (255, 255))   # Silently converts to FP32
        img = tf.image.convert_image_dtype(img, tf.float32)   # this operation is skipped silently
```

**Output:**

```bash
$ python main.py 
Processing with the correct order of operations ... No warning shall be printed ...

Processing with the dangerous order of operations ... Only one warning shall be printed ...
WARNING:tensorflow:The operation `_resize_images_common` called by `main.py:L24` has silently converted the data type from `<dtype: 'uint8'>` to `<dtype: 'float32'>`.
This might have changed the image format in ways that make it incompatible with other tf.image functions.
```

**Comment by gbaned** [comment]
@martinwicke Can you please take a look on above comments from @DEKHTIARJonathan. Thanks!

**Comment by gbaned** [comment]
@martinwicke Can you please assist on above comments from @DEKHTIARJonathan. Thanks!

**Review by alextp — APPROVED** [review]

**Comment by gbaned** [comment]
@DEKHTIARJonathan  Can you please check build failures. Thanks!

**Comment by gbaned** [comment]
@DEKHTIARJonathan, Any update on this PR? Please. Thanks!

**Comment by DEKHTIARJonathan** [comment]
I need a bit more time

**Comment by gbaned** [comment]
@DEKHTIARJonathan, Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
@DEKHTIARJonathan, Any update on this PR? Please. Thanks!

**Comment by DEKHTIARJonathan** [comment]
@gbaned sorry I need more time. I can't focus on this atm.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
@DEKHTIARJonathan, Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
